src/rpc/blockchain.cpp

```diff
@@ -184,7 +184,7 @@ UniValue blockToJSON(BlockManager& blockman, const CBlock& block, const CBlockIn
         case TxVerbosity::SHOW_DETAILS_AND_PREVOUT:
             CBlockUndo blockUndo;
             const bool is_not_pruned{WITH_LOCK(::cs_main, return !blockman.IsBlockPruned(blockindex))};
-            const bool have_undo{is_not_pruned && blockman.UndoReadFromDisk(blockUndo, blockindex)};
+            const bool have_undo{is_not_pruned && WITH_LOCK(::cs_main, return blockindex.nStatus & BLOCK_HAVE_UNDO) && blockman.UndoReadFromDisk(blockUndo, blockindex)};
```

Comment:
In 041b78ccede9974117dac5d8623fedec2f32471d:

nit: I think it would be cleaner and easier to read if you break this up.

```suggestion
            const bool have_undo{is_not_pruned && WITH_LOCK(::cs_main, return blockindex.nStatus & BLOCK_HAVE_UNDO)};
            if (have_undo) {
                blockman.UndoReadFromDisk(blockUndo, blockindex);
            }
```

(same in `rpc/rawtransaction.cpp`)

Reply:
I can break this up, but I think the suggestion doesn't work because we still need to feed the result into `have_undo` so that if `UndoReadFromDisk` fails for an unexpected reason, we don't attempt to access `blockUndo` below and segfault.

Reply:
> I think the suggestion doesn't work because we still need to feed the result into have_undo so that if UndoReadFromDisk fails for an unexpected reason like disk errors, we don't attempt to access blockUndo below and segfault.

I see, I think in that scenario it would be best to throw an error right there instead of ignoring it silently and giving the user an incomplete response? I mean, we have all indication that we should have the undo data in that scenario and then we can not read it, it seems like an error that the user should know about.

> In rpc/rawtransaction.cpp, did you mean to introduce a variable like skip_details or something to break this up?

I think that would work, I don't remember if I had something explicit in mind when I wrote it. It's just an awfully long list of conditions already and summing parts of them up in a variable with a concise name would make it easier to parse. I have left two additional suggestions there for shortening it inline and I think there it's the same question if we really want to silently return a limited result if we fail to read from disk what we are supposed to have. I'll wait for you feedback on that part and my inline suggestions. If nothing else I think it would also help a lot to just add line breaks after every condition.

Reply:
I think it makes sense to throw an error in this situation. Although it's not necessarily due to disk corruption, it would also be possible that the block/undo data is pruning event right after the blockindex check (`cs_main` is released) and before the `UndoReadFromDisk` call.

We also have a functional test for just this situation that wouldn't work anymore ([here](https://github.com/bitcoin/bitcoin/blob/c66c68345efb0bb3d5613ebac703cde779fa0f01/test/functional/rpc_blockchain.py#L615-L628)) but that tests doesn't actually prune a block but deletes an undo file to simulate pruning. So we'd either have to just remove it or actually prune a block... I'll try this out in the next days.